docs: Add Bitbucket Pipelines configuration (#1402)
testcontainers/testcontainers-dotnet: docs/cicd/index.md

variables:
  DOCKER_HOST: tcp://docker:2375
```

> ## Bitbucket CI/CD
> 
> Enable your bitbucket CI/CD pipeline as usual on the admin-pipelines-settings page. After enabling your pipeline, replace the content of *bitbucket-pipelines.yml* (you can find this file in the root of your source folder) with:

Reviewer: ```suggestion
## Bitbucket Pipelines

Enable Bitbucket Pipelines as usual on the **Repository settings → Pipelines → Settings** page. After enabling your pipeline, replace the contents of the `bitbucket-pipelines.yml` file, located at the root of your repository, with the following:
```

I guess this only works for the cloud version, not for the on-premise (Bamboo) setup, right?

Author: Good question, I only have the cloud version available to test (repo link above).

```yml

> image: mcr.microsoft.com/dotnet/sdk:9.0
> 
> options:
>   docker: true
> 
> pipelines:
>   default:
>     - step:
>         name: Build and Test
>         services:
>           - docker
>         caches:
>           - dotnetcore
>         script:
>           - dotnet restore
>           - dotnet build --configuration Release
>           - dotnet test --configuration Release --no-build --logger "trx;LogFileName=test-results.trx"
>         artifacts:
>           - test-results.trx

Reviewer: ```suggestion
image: mcr.microsoft.com/dotnet/sdk:8.0
options:
  docker: true
pipelines:
  default:
    - step:
      script:
        - dotnet test
      services:
        - docker
      environment:
        # Bitbucket Pipelines does not support Ryuk:
        # https://dotnet.testcontainers.org/api/resource_reaper/.
        TESTCONTAINERS_RYUK_DISABLED: true
```

I would like to include only the minimal necessary configuration. I believe the cache, restore, and build steps are not essential and depend on how individual developers or teams set up their CI. I expect the following steps are sufficient to enable Testcontainers support:

1. Enable the Docker option by setting it to `true`.
2. Use the Docker service.
3. Disable Ryuk.

WDYT?

Author: Very reasonable!

I've setup a repo with an example pipeline at https://bitbucket.org/xenopz-public/bitbucket-pipeline/ with the original code. I'll add your environment variable asap.

Author: I'm afraid I get the following error:

```
  Error Message:

   Docker.DotNet.DockerApiException : Docker API responded with status code=Forbidden, response={"message":"authorization denied by plugin pipelines: --mounts is not allowed"}
```

Reviewer: Are you sure the environment variable is properly set? It seems like it isn't.

Author: Only works if I add the env variable under the repository variables, which is fine if documented.

pipeline test: https://bitbucket.org/xenopz-public/bitbucket-pipeline/pipelines/results/9/steps/%7B62226a80-81f4-43b9-96c2-e5ca538e7462%7D

bitbucket pipeline: https://bitbucket.org/xenopz-public/bitbucket-pipeline/src/main/bitbucket-pipelines.yml

Reviewer: It sounds a bit unusual, but I don't know anything about Bitbucket Pipelines. Have you tried setting it up as described [here](https://community.atlassian.com/forums/Bitbucket-questions/Setting-env-vars-in-pipeline-in-configuration-file/qaq-p/1776026#M65175)?

```yml
script:
  - dotnet test
services:
  - docker
variables:
  # Bitbucket Pipelines does not support Ryuk:
  # https://dotnet.testcontainers.org/api/resource_reaper/.
  TESTCONTAINERS_RYUK_DISABLED: true
```

or

```yml
script:
  - export TESTCONTAINERS_RYUK_DISABLED=true
  - dotnet test
services:
  - docker
```

Otherwise, we will follow your suggested approach and document it accordingly. Thanks, for the tests and feedback.

Author: sure, the export syntax works, the variables syntax does not work

the last pipeline build in my repo is with the variables syntax, so you can double check if you want to (unfortunately, I am also not an expert in Bitbucket Pipelines.), see https://bitbucket.org/xenopz-public/bitbucket-pipeline/pipelines/results/11/steps/%7B500f46ab-62c9-4a08-9ac2-d67598fb7e9a%7D

to resume, two options work, repository variables or export var=value

Reviewer: I checked the documentation for the implementations in other languages, and for example, Golang also uses the export variation in the script step. I believe we should keep this approach for the sake of alignment: https://golang.testcontainers.org/system_requirements/ci/bitbucket_pipelines/.

Author: reasonable!

```

> 
> In your code, you can use a [standard hello world xUnit application](https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-with-dotnet-test). Replace the test code with the following code:
> 
> ```csharp
> [Fact]
> public async Task Test_HelloWorldContainer()
> {
>     //https://github.com/testcontainers/testcontainers-dotnet/issues/492
>     TestcontainersSettings.ResourceReaperEnabled = false;
>     //Arrange
>     
>     // Create a new instance of a container.
>     var container = new ContainerBuilder()
>       // Set the image for the container to "testcontainers/helloworld:1.1.0".
>       .WithImage("testcontainers/helloworld:1.1.0")
>       // Bind port 8080 of the container to a random port on the host.
>       .WithPortBinding(8080, true)
>       // Wait until the HTTP endpoint of the container is available.
>       .WithWaitStrategy(Wait.ForUnixContainer().UntilHttpRequestIsSucceeded(r => r.ForPort(8080)))
>       // Build the container configuration.
>       .Build();
>     
>     await container.StartAsync();
> 
>     // Act
>     var httpClient = new HttpClient
>     {
>         BaseAddress = new Uri($"http://{container.Hostname}:{container.GetMappedPublicPort(8080)}")
>     };
> 
>     var response = await httpClient.GetAsync("/");
>     var content = await response.Content.ReadAsStringAsync();
> 
>     // Assert
>     Assert.True(response.IsSuccessStatusCode, "Expected a successful HTTP response");
>     Assert.Contains("Hello world", content, StringComparison.OrdinalIgnoreCase);
> }
> ```
> 
> Note the *Reaper* config:
> 
> ```csharp
> TestcontainersSettings.ResourceReaperEnabled = false;
> ```
> 
> If you don't enable this config, the test container will not function correctly.
> 
> Run the pipeline by pushing code or by clicking *Run pipeline* on the Pipelines page.

Reviewer: ```suggestion
```

With the suggested setup above, ideally no additional configuration in the source code is necessary. I expect Testcontainers to run immediately without requiring any example code here. The regular example on the start page should work as is.

Author: The regular example is used in my test code (with xUnit - see repo above) and works correctly.
